external-crates/move/testing-infra/move-proc-macros/src/lib.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Data, DataEnum, DeriveInput};
+
+/// This proc macro generates a function `order_to_variant_map` which returns a map
+/// of the position of each variant to the name of the variant.
+/// It is intended to catch changes in enum order when backward compat is required.
+/// ```rust,ignore
+///    /// Example for this enum
+///    #[derive(EnumVariantOrder)]
+///    pub enum MyEnum {
+///         A,
+///         B(u64),
+///         C{x: bool, y: i8},
+///     }
+///     let order_map = MyEnum::order_to_variant_map();
+///     assert!(order_map.get(0).unwrap() == "A");
+///     assert!(order_map.get(1).unwrap() == "B");
+///     assert!(order_map.get(2).unwrap() == "C");
+/// ```
+#[proc_macro_derive(EnumVariantOrder)]
```

Comment:
Instead of a deriving macro + adding tests, can we roll it into one attribute macro, that accepts an optional parameter for the filename to store the snapshot?, i.e. something like:

```
#[test_variant_order("unit_tests/enums/my_enum.yaml")]
pub enum MyEnum {
    A,
    B(u64),
    C { x: bool, y: i8 },
}
```

The benefits are that we can generate the `#[test]` as well, so it won't get forgotten, and we can also gate the generation of the variant ordering to only be available in `#[test]` mode, to avoid having to build it for release, and/or introduce it as an option in autocompletes, etc.